store/engine/embedded/raft.go

```diff
@@ -0,0 +1,550 @@
+ 
+	 rpeers := make([]raft.Peer, len(rc.peers))
+	 for i := range rpeers {
+		 rpeers[i] = raft.Peer{ID: uint64(i + 1)}
+	 }
+	 c := &raft.Config{
+		 ID:                        uint64(rc.id),
+		 ElectionTick:              10,
+		 HeartbeatTick:             1,
+		 Storage:                   rc.raftStorage,
+		 MaxSizePerMsg:             1024 * 1024,
+		 MaxInflightMsgs:           256,
+		 MaxUncommittedEntriesSize: 1 << 30,
+	 }
+ 
+	 if oldwal || rc.join {
+		 rc.node = raft.RestartNode(c)
+	 } else {
+		 rc.node = raft.StartNode(c, rpeers)
+	 }
+ 
+	 rc.transport = &rafthttp.Transport{
+		 Logger:      logger.Get(),
+		 ID:          types.ID(rc.id),
+		 ClusterID:   0x1000,
+		 Raft:        rc,
+		 ServerStats: stats.NewServerStats("", ""),
+		 LeaderStats: stats.NewLeaderStats(zap.NewExample(), strconv.Itoa(rc.id)),
+		 ErrorC:      make(chan error),
+ 
+		 DialRetryFrequency: 1,
+	 }
+ 
+	 if err := rc.transport.Start(); err != nil {
+		 logger.Get().With(zap.Error(err)).Panic("Failed to start raft http server")
+	 }
+	 for i := range rc.peers {
+		 if i+1 != rc.id {
+			 rc.transport.AddPeer(types.ID(i+1), []string{rc.peers[i]})
+		 }
+	 }
+ 
+	 go rc.serveRaft()
+	 go rc.serveChannels()
+ }
+ 
+ // stop closes http, closes all channels, and stops raft.
+ func (rc *raftNode) stop() {
+	 rc.stopHTTP()
+	 close(rc.commitC)
+	 close(rc.errorC)
+	 rc.node.Stop()
+ }
+ 
+ func (rc *raftNode) stopHTTP() {
+	 rc.transport.Stop()
+	 close(rc.httpStopCh)
+	 <-rc.httpDoneCh
+ }
+ 
+ func (rc *raftNode) publishSnapshot(snapshotToSave raftpb.Snapshot) {
+	 if raft.IsEmptySnap(snapshotToSave) {
+		 return
+	 }
+ 
+	 logger.Get().Sugar().Infof("Publishing snapshot at index %d", rc.snapshotIndex)
+	 defer logger.Get().Sugar().Infof("Finished publishing snapshot at index %d", rc.snapshotIndex)
+ 
+	 if snapshotToSave.Metadata.Index <= rc.appliedIndex {
+		 logger.Get().Sugar().Fatalf("snapshot index [%d] should > progress.appliedIndex [%d]", snapshotToSave.Metadata.Index, rc.appliedIndex)
+	 }
+	 rc.commitC <- nil // trigger kvstore to load snapshot
+ 
+	 rc.confState = snapshotToSave.Metadata.ConfState
+	 rc.snapshotIndex = snapshotToSave.Metadata.Index
+	 rc.appliedIndex = snapshotToSave.Metadata.Index
+ }
+ 
+ var snapshotCatchUpEntriesN uint64 = 10000
+ 
+ func (rc *raftNode) maybeTriggerSnapshot(applyDoneC <-chan struct{}) {
+	 if rc.appliedIndex-rc.snapshotIndex <= rc.snapCount {
+		 return
+	 }
+ 
+	 // wait until all committed entries are applied (or server is closed)
+	 if applyDoneC != nil {
+		 select {
+		 case <-applyDoneC:
+		 case <-rc.stopCh:
+			 return
+		 }
+	 }
+ 
+	 logger.Get().Sugar().Infof("start snapshot [applied index: %d | last snapshot index: %d]", rc.appliedIndex, rc.snapshotIndex)
+	 data, err := rc.getSnapshot()
+	 if err != nil {
+		 logger.Get().With(zap.Error(err)).Panic("Failed to snapshot")
+	 }
+	 snapshot, err := rc.raftStorage.CreateSnapshot(rc.appliedIndex, &rc.confState, data)
+	 if err != nil {
+		 panic(err)
```

Comment:
Why do we use panic here and followings?

Reply:
If error happens when snapshotting, I believe it is an unrecoverable one. Keeping running further does no good.